### ISSUE TITLE
[clang-pseudo] Forest.h - don't inherit from std::iterator

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#57005]
Now that we've updated to C++17 MSVC gives very verbose warnings about not creating classes that inherit from std::iterator:
```
Building CXX object tools\clang\tools\extra\pseudo\tool\CMakeFiles\clang-pseudo.dir\HTMLForest.cpp.obj

E:\llvm\llvm-project\clang-tools-extra\pseudo\include\clang-pseudo/Forest.h(202): warning C4996: 'std::iterator<std::input_iterator_tag,const clang::pseudo::ForestNode,ptrdiff_t,const clang::pseudo::ForestNode *,const clang::pseudo::ForestNode &>': warning STL4015: The std::iterator class template (used as a base class to provide typedefs) is deprecated in C++17. (The <iterator> header is NOT deprecated.) The C++ Standard has never required user-defined iterators to derive from std::iterator. To fix this warning, stop deriving from std::iterator and start providing publicly accessible typedefs named iterator_category, value_type, difference_type, pointer, and reference. Note that value_type is required to be non-const, even for constant iterators.
```
while MSVC does give us the option of silencing these warnings via _SILENCE_CXX17_RESULT_OF_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS, I'd much prefer to see these fixed properly.

**Comment by RKSimon** [comment]
This silences the warning on MSVC, haven't tested other compilers though:
```
class ForestNode::RecursiveIterator {
  llvm::DenseSet<const ForestNode *> Seen;
  struct StackFrame {
    const ForestNode *Parent;
    unsigned ChildIndex;
  };
  std::vector<StackFrame> Stack;
  const ForestNode *Cur;

public:
  using iterator_category = std::input_iterator_tag;
  using value_type = ForestNode;
  using difference_type = ptrdiff_t;
  using pointer = const ForestNode*;
  using reference = const ForestNode&;

  RecursiveIterator(const ForestNode *N = nullptr) : Cur(N) {}

  const ForestNode &operator*() const { return *Cur; };
  void operator++();
  bool operator==(const RecursiveIterator &I) const { return Cur == I.Cur; }
  bool operator!=(const RecursiveIterator &I) const { return !(*this == I); }
};
```

**Comment by nikic** [comment]
LLVM provides iterator_facade_base for this purpose.
